src/main/java/org/zalando/nakadi/service/EventStream.java

```diff
@@ -159,13 +183,53 @@ public static String createStreamEvent(final Cursor cursor, final List<String> e
 
     private void sendBatch(final NakadiCursor topicPosition, final List<String> currentBatch)
             throws IOException {
-        // create stream event batch for current partition and send it; if there were
-        // no events, it will be just a keep-alive
-        final String streamEvent = createStreamEvent(cursorConverter.convert(topicPosition), currentBatch);
-        final byte[] batchBytes = streamEvent.getBytes(UTF8);
-        outputStream.write(batchBytes);
-        bytesFlushedMeter.mark(batchBytes.length);
-        outputStream.flush();
+        if(featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.SEND_BATCH_VIA_OUTPUT_STREAM)) {
+            writeStreamEvent(outputStream, cursorConverter.convert(topicPosition), currentBatch);
+        } else {
+            // create stream event batch for current partition and send it; if there were
+            // no events, it will be just a keep-alive
+            final String streamEvent = createStreamEvent(cursorConverter.convert(topicPosition), currentBatch);
+            final byte[] batchBytes = streamEvent.getBytes(UTF8);
+            outputStream.write(batchBytes);
+            bytesFlushedMeter.mark(batchBytes.length);
+            outputStream.flush();
+        }
+    }
+
+    @VisibleForTesting
+    void writeStreamEvent(final OutputStream os, final Cursor cursor, final List<String> events)
+        throws IOException {
+
+        int byteCount = B_FIXED_BYTE_COUNT;
+
+        os.write(B_CURSOR_PARTITION_BEGIN);
+        final byte[] partition = cursor.getPartition().getBytes(StandardCharsets.UTF_8);
+        os.write(partition);
+        byteCount += partition.length;
+        os.write(B_OFFSET_BEGIN);
+        final byte[] offset = cursor.getOffset().getBytes(StandardCharsets.UTF_8);
+        os.write(offset);
+        byteCount += offset.length;
+        os.write(B_CURSOR_PARTITION_END);
+        if (!events.isEmpty()) {
+            os.write(B_EVENTS_ARRAY_BEGIN);
+            for (int i = 0; i < events.size(); i++) {
+                final byte[] event = events.get(i).getBytes(StandardCharsets.UTF_8);
```

Comment:
@lhotari 

> Perhaps consuming byte[] instead of String is something that would be worth doing?

It would have been better to have coordinated first before spinning up #605; see https://github.com/zalando/nakadi/pull/589#issuecomment-285317114 (relative to internal work this is part of a ongoing tuning exercise). I'll starting commenting on

I'm aware of CountingOutputStream; imvho it would be better to do that as part of a separate pr to keep the changes isolated (count bytes isn't ideal but it follows what's already here).